src/codegen/transformations.lisp

```diff
@@ -11,204 +11,159 @@
    (#:tc #:coalton-impl/typechecker))
   (:export
    #:traverse
-   #:traverse-bindings
+   #:traverse-with-binding-list
    #:update-function-env
-   #:make-function-t))
+   #:make-function-t
+   #:substitute-fresh-type-variables))
 
 (in-package #:coalton-impl/codegen/transformations)
 
-(defun traverse-bindings (bindings funs)
-  (declare (type binding-list bindings)
-           (values binding-list))
-  (loop :for (name . node) :in bindings
-        :collect (cons name (traverse node funs nil))))
+(defun call-if (key funs args node)
+  (declare (type symbol key)
+           (type list   funs)
+           (type list   args)
+           (type node   node)
+           (values node &optional))
+  (or
+   (alexandria:when-let ((f (cdr (assoc key funs))))
+     (apply f (cons node args)))
+   node))
 
-(defun call-if (node key funs bound-variables)
+(defmacro traverse-if (key funs args before-node else-expr)
+  (let ((before-key
+          (alexandria:make-keyword (concatenate 'string "BEFORE-" (symbol-name (cadr key)))))
```

Comment:
If we quote the keywords, then the macro needs their symbol from `cadr`

Reply:
In this case the keywords are syntax, not data.

Reply:
Ok, for now I reverted the keywords being passed to this macro to unquoted